src/storage/kv/transaction.rs

```diff
@@ -459,6 +518,84 @@ impl Transaction {
         self.read_set.lock().clear();
         self.snapshot.take();
     }
+
+    /// After calling this method the subsequent modifications within this
+    /// transaction can be rolled back by calling [`rollback_to_savepoint`].
+    ///
+    /// This method is stackable and can be called multiple times with the
+    /// corresponding calls to [`rollback_to_savepoint`].
+    ///
+    /// [`rollback_to_savepoint`]: Transaction::rollback_to_savepoint
+    pub fn set_savepoint(&mut self) -> Result<()> {
+        // If the transaction mode is not mutable (i.e., it's read-only), return an error.
+        if !self.mode.mutable() {
+            return Err(Error::TransactionReadOnly);
+        }
+        // If the transaction is closed, return an error.
+        if self.closed {
+            return Err(Error::TransactionClosed);
+        }
+
+        // Bump the latest savepoint number.
+        self.savepoints += 1;
+
+        Ok(())
+    }
+
+    /// Rollback the state of the transaction to the latest savepoint set by
+    /// calling [`set_savepoint`].
+    ///
+    /// [`set_savepoint`]: Transaction::set_savepoint
+    pub fn rollback_to_savepoint(&mut self) -> Result<()> {
+        // If the transaction mode is not mutable (i.e., it's read-only), return an error.
+        if !self.mode.mutable() {
+            return Err(Error::TransactionReadOnly);
+        }
+        // If the transaction is closed, return an error.
+        if self.closed {
+            return Err(Error::TransactionClosed);
+        }
+        // Check that the savepoint is set
+        if self.savepoints == 0 {
+            return Err(Error::TransactionWithoutSavepoint);
+        }
+
+        // Remove entries marked for rollback since the last call
+        // to set_savepoint() from the write set, the write order map
+        // and the snapshot.
+        self.write_set.retain(|entry| {
+            if entry.savepoint_no == self.savepoints {
+                let hashed_key = sha256(entry.e.key.clone());
+                self.write_order_map.remove(&hashed_key);
+                self.snapshot
```

Comment:
There could be a potential issue here when the same key is overwritten. Let's say we have

```
    #[tokio::test]
    async fn foo() {
        // Scan set is only considered for SSI.
        let (store, _) = create_store(true);

        let k1 = Bytes::from("k1");
        let value = Bytes::from("test_value1");
        let value2 = Bytes::from("test_value2");

        let mut txn1 = store.begin().unwrap();
        txn1.set(&k1, &value).unwrap();
        txn1.set_savepoint().unwrap();
        txn1.set(&k1, &value2).unwrap();
        txn1.rollback_to_savepoint().unwrap();
        let val = txn1.get(&k1).unwrap();
        assert!(val.is_some());
    }
```

Because we are deleting the entire key from the snapshot (which does not track the savepoint number), this value is deleted, and could become an issue. The same would apply for scans too because scans also read from the snapshot